google/cloud/internal/async_streaming_write_rpc_auth_test.cc

```diff
@@ -0,0 +1,178 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/async_streaming_write_rpc_auth.h"
+#include "google/cloud/completion_queue.h"
+#include "google/cloud/testing_util/mock_grpc_authentication_strategy.h"
+#include "google/cloud/testing_util/status_matchers.h"
+#include <gmock/gmock.h>
+#include <memory>
+#include <string>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+namespace {
+
+using ::google::cloud::testing_util::IsOk;
+using ::google::cloud::testing_util::MockAuthenticationStrategy;
+using ::google::cloud::testing_util::StatusIs;
+using ::testing::Contains;
+using ::testing::Pair;
+using ::testing::StrictMock;
+
+struct FakeRequest {
+  std::string key;
+};
+
+struct FakeResponse {
+  std::string key;
+  std::string value;
+};
+
+using BaseStream = AsyncStreamingWriteRpc<FakeRequest, FakeResponse>;
+using AuthStream = AsyncStreamingWriteRpcAuth<FakeRequest, FakeResponse>;
+
+class MockStream : public BaseStream {
+ public:
+  MOCK_METHOD(void, Cancel, (), (override));
+  MOCK_METHOD(future<bool>, Start, (), (override));
+  MOCK_METHOD(future<bool>, Write, (FakeRequest const&, grpc::WriteOptions),
+              (override));
+  MOCK_METHOD(future<bool>, WritesDone, (), (override));
+  MOCK_METHOD(future<StatusOr<FakeResponse>>, Finish, (), (override));
+  MOCK_METHOD(StreamingRpcMetadata, GetRequestMetadata, (), (const, override));
+};
+
+TEST(AsyncStreamingWriteRpcAuth, Start) {
+  auto factory = AuthStream::StreamFactory([](std::unique_ptr<
+                                               grpc::ClientContext>) {
+    auto mock = absl::make_unique<StrictMock<MockStream>>();
+    EXPECT_CALL(*mock, Start).WillOnce([] { return make_ready_future(true); });
+    EXPECT_CALL(*mock, Write).WillOnce([] { return make_ready_future(false); });
```

Comment:
optional nit: have this return true, and call `uut->WritesDone()` after, just to test everything.

Reply:
Done